webapp/src/ts/services/training-cards.service.ts

```diff
@@ -0,0 +1,137 @@
+import { Injectable } from '@angular/core';
+import { Store } from '@ngrx/store';
+
+import { XmlFormsService } from '@mm-services/xml-forms.service';
+import { TrainingCardsComponent } from '@mm-modals/training-cards/training-cards.component';
+import { DbService } from '@mm-services/db.service';
+import { GlobalActions } from '@mm-actions/global';
+import { ModalService } from '@mm-modals/mm-modal/mm-modal';
+import { SessionService } from '@mm-services/session.service';
+import { RouteSnapshotService } from '@mm-services/route-snapshot.service';
+
+export const TRAINING_PREFIX: string = 'training';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TrainingCardsService {
+  private globalActions;
+
+  constructor(
+    private store: Store,
+    private xmlFormsService: XmlFormsService,
+    private dbService: DbService,
+    private modalService: ModalService,
+    private sessionService: SessionService,
+    private routeSnapshotService: RouteSnapshotService,
+  ) {
+    this.globalActions = new GlobalActions(store);
+  }
+
+  private getAvailableTrainingForms(xForms, userCtx) {
+    const today = new Date();
+
+    return xForms
+      .map(xForm => ({
+        id: xForm._id,
+        code: xForm.internalId,
+        startDate: xForm.context?.start_date,
+        duration: xForm.context?.duration,
+        userRoles: xForm.context?.user_roles,
+      }))
+      .filter(form => {
+        const hasRole = form.userRoles?.find(role => userCtx.roles.includes(role));
+        if (!hasRole) {
+          return false;
+        }
+
+        const startDate = new Date(form.startDate);
+        if (startDate > today) {
+          return false; // Training has not started yet
+        }
+
+        if (!form.duration) {
+          return true; // Training never ends
+        }
+
+        const endDate = new Date(form.startDate);
+        endDate.setDate(endDate.getDate() + form.duration);
+        return endDate > today;
+      });
+  }
+
+  private async getCompletedTrainings(userCtx) {
+    const docs = await this.dbService
+      .get()
```

Comment:
My first guess for adding them to `medic` would be that maybe we want to use them to generate tasks at some point, or perhaps adding sentinel transitions. I know that displaying the training as a report was under debate and was considered a desirable outcome. 

I understand the point of adding docs to secondary databases. I think that, for a high volume of docs that would be justified. I don't expect users to generate thousands of trainings. Adding multiple databases adds some complexity to the code, because we want these databases to sync, and will probably need to add some restrictions about which docs we allow users to sync (fe don't allow a user to submit a training for someone else!). Speaking of this, @latin-panda do we have any protection around users syncing training records for other users? 

Adding the docs in users meta is a valid point tho.

Reply:
Initially, we tried to make the Training Cards as a particular type of document in CouchDB (different from `data-record`), but to replicate these, we would have to update the views; the indexing runs from the start and not just from new documents, so it is too expensive for projects. After a conversation with @garethbowen, we decided to make Training Cards like a regular report and inherit much of what reports offer. For example, there is some interest in building targets or tasks based on Training Cards.

> do we have any protection around users syncing training records for other users?

When it comes to syncing, it just behaves as regular reports. Projects might want to let the supervisors check if CHWs are completing the training. 

However, we know if someone has completed the training or not. So, for example: 
`Anna,` a CHW from an area, completed the training and sync.
`Zoe,` a CHW from the area's parent facility, has not completed the training yet; then she syncs the app and can see that `Anna` has submitted the training. The app will still display the modal with the training to `Zoe` because she needs to complete it and the training targets her user role.

CC: @michaelkohn

Reply:
Maybe worth mentioning that this is MVP; the feature can do more or do some things differently than we can develop based on the project's feedback/interests. Such as having a dedicated page to visualize pending trainings - it has been the most popular request until now.

Reply:
> there is some interest in building targets or tasks based on Training Cards.

Ah, this makes a lot of sense!  So, tasks/targets can only be based on content from the `medic` db?  I guess that keeps things as straight-forward as possible....